### ISSUE TITLE
[Apache]Libyami decoder interface decode() will block forever after consumed all internal surfaces 

### ISSUE THREAD

**Issue by mypopydev** [01org/libyami#493]
In my mind,  **decode()/getOutPut()** is not block function ever we can't decode sucess/get a full decoded frame, and we need to decoupling the decode()/getOutput().
1.  **decode()** to give the decoder raw compressed data(queueing) and return is no-blocking.
2. After decoding, call **getOutput()**. On success, it will return an uncompressed video data and release internal surface resource.

And it's working like ffmpeg push/pull interface  **avcodec_send_packet()/avcodec_receive_frame()**  https://libav.org/documentation/doxygen/master/group__lavc__encdec.html

But now, decode() function will blocking forever after consumed all internal surfaces, and the caller can't find any recover way.

Basically, used 

`while(1) {
     status = decoder->decode();
     if (status == SUCCESS)
        do something
     if (status == FAIL or EAGAIN)
        try again later
 }`

will reproduce this issue. I think ever we consumed all internal surfaces, we need to return a error like  EAGAIN (Resource temporarily unavailable; the call might work if you try again later. The macro EWOULDBLOCK is another name for EAGAIN; they are always the same in the GNU C Library.), Blocking in decode() is a bad action for the caller.

**Comment by xuguangxin** [comment]
Yes, for historic reason, we use the blocking behavior in v4l2decoder. 

> https://github.com/01org/libyami/blob/apache/v4l2/v4l2_codecbase.cpp#L375

If we change this, it will break v4l2decoder. It may need more time to change v4l2decoder related things.
One of work around is never make decode blocking, always make sure you have enough surface before you call decode.
Our allocator will allocate needed surface + an extra size (5 in current design). This means if you hold less than 5 frames in hand, you never been blocked in decode. 
Other suggestion is you do not have to depend on default extra size; you should implement your own allocator and set it to 

> https://github.com/01org/libyami/blob/apache/interface/VideoDecoderInterface.h#L61.

Then you can allocate as many surface as you want.

**Comment by mypopydev** [comment]
I think clean the v4l2decoder/internal surface pool management  is the right way. :). And I know **releaseLock()** have maked with deprecated.

BTW: I guess some caller don't want to touch the external allocator :)

**Comment by xuguangxin** [comment]
fix by fbbafc8
